### PR TITLE
ceph: mock object store user create

### PR DESCRIPTION
**Pull request by leseb**
**Description of your changes:**

Since the mock client was added to the rgw/admin package from the
go-ceph project in https://github.com/ceph/go-ceph/pull/532 we can now
mock the HTTP client and validate the entire reconcile.

Signed-off-by: Sébastien Han <seb@redhat.com>

**Which issue is resolved by this Pull Request:**
Resolves #

**Checklist:**

- [ ] **Commit Message Formatting**: Commit titles and messages follow guidelines in the [developer guide](https://rook.io/docs/rook/master/development-flow.html#commit-structure).
- [ ] **Skip Tests for Docs**: Add the flag for skipping the build if this is only a documentation change. See [here](https://github.com/rook/rook/blob/master/INSTALL.md#skip-ci) for the flag.
- [ ] **Skip Unrelated Tests**: Add a flag to run tests for a specific storage provider. See [test options](https://github.com/rook/rook/blob/master/INSTALL.md#test-storage-provider).
- [ ] Reviewed the developer guide on [Submitting a Pull Request](https://rook.io/docs/rook/master/development-flow.html#submitting-a-pull-request)
- [ ] Documentation has been updated, if necessary.
- [ ] Unit tests have been added, if necessary.
- [ ] Integration tests have been added, if necessary.
- [ ] Pending release notes updated with breaking and/or notable changes, if necessary.
- [ ] Upgrade from previous release is tested and upgrade user guide is updated, if necessary.
- [ ] Code generation (`make codegen`) has been run to update object specifications, if necessary.

### CODE REVIEW

**Review by travisn — COMMENTED** [review]
LGTM, but the smoke suite seems to have a failure

File: pkg/operator/ceph/object/dependents_test.go [hunk 1]
```diff
@@ -0,0 +1,242 @@
+/*
```
Comment (leseb) [inline pkg/operator/ceph/object/dependents_test.go:1]: @BlaineEXE PTAL at this file 🙂

**Comment by travisn** [comment]
The object store user test is failing in the smoke suite, the operator log shows:
```
2021-07-21 17:19:04.357132 I | ceph-object-store-user-controller: creating ceph object user "rook-user" in namespace "smoke-ns"
2021-07-21 17:19:04.404559 E | ceph-object-store-user-controller: failed to reconcile failed to create/update object store user "rook-user": failed to get details from ceph object user "rook-user": Get "http://rook-ceph-rgw-teststore.smoke-ns.svc:80/admin/user?display-name=A%20rook%20RGW%20user&format=json&uid=rook-user": dial tcp: lookup rook-ceph-rgw-teststore.smoke-ns.svc on 10.96.0.10:53: no such host
```

**Review by travisn — APPROVED** [review]
The CI is happy now.

File: pkg/operator/ceph/object/user/controller.go [hunk 2]
```diff
@@ -314,11 +318,15 @@ func (r *ReconcileObjectStoreUser) initializeObjectStoreContext(u *cephv1.CephOb
...
-	if err != nil {
-		return errors.Wrap(err, "failed to initialized rgw admin ops client api")
+	// This is needed for the unit tests so the mock client does not get overridden
+	// This should not impact normal operations in production
+	if !IsRunningInUnitTest {
+		opsContext, err := object.NewMultisiteAdminOpsContext(objContext, &store.Spec)
+		if err != nil {
+			return errors.Wrap(err, "failed to initialized rgw admin ops client api")
+		}
+		r.objContext = opsContext
```
Comment (BlaineEXE) [inline pkg/operator/ceph/object/user/controller.go:328]: A pattern we've used in the past that seems to me like it makes for slightly better-looking code is to be able to replace the `object.NewMultisiteAdminOpsContext()` for unit tests like I outline below. I think this implementation works, and I think we can merge it, but let me know if you like the below suggestion better.

```go
var newMultisiteAdminOpsCtxFunc = object.NewMultisiteAdminOpsContext

func main() {
    // ...
    opsCtx, err := newMultisiteAdminOpsCtxFunc(objContext, &store.Spec)
    // ...
}

---

func TestMain(t *testing.T) {
    newMultisiteAdminOpsCtxFunc = func(/* args */) /* returns */ {
        return /* a mock ctx */
    }
}
```

**Review by BlaineEXE — COMMENTED** [review]
Looks good to me. Thanks for doing all this. I just had one comment thought.

File: pkg/operator/ceph/object/user/controller.go (hunk 2 above)
Comment (leseb) [reply pkg/operator/ceph/object/user/controller.go:328]: I like this option better, much more elegant. Thanks!

**Review by BlaineEXE — APPROVED** [review]

**Pull request merged**